### PR TITLE
🔴 [VLM] modeling updates

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

I forgot to change a few more things weren't changed in previous PRs. Specifically:
- Attention refactor of Qwen and Persimmon (for Fuyu)
- The helper `get_image_features` has to return images of certain shape and format. May be breaking and we had the helper as public prev 🙈

### CODE REVIEW

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -384,7 +384,7 @@ def forward(
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
-        **kwargs,  # NOOP kwarg for now
+        **kwargs: Unpack[FlashAttentionKwargs],
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma/modeling_gemma.py:387]: Was added for GradAcc issues in `git blame`, I believe we should have fixed grad acc already no?

File: src/transformers/models/idefics3/modeling_idefics3.py [hunk 2]
```diff
@@ -742,7 +744,7 @@ def get_image_features(self, pixel_values: torch.FloatTensor, pixel_attention_ma
 
         # Modality projection & resampling
         image_hidden_states = self.connector(image_hidden_states.last_hidden_state)
-        image_hidden_states = image_hidden_states.view(-1, image_hidden_states.shape[-1])
+        image_hidden_states = image_hidden_states.view(batch_size, -1, image_hidden_states.shape[-1])
```
Comment (zucchini-nlp) [inline src/transformers/models/idefics3/modeling_idefics3.py:747]: not breaking because `get_image_features` didn't exist before. I added it newly and now changing shape to make vLLM happy

**Comment by Tcc0403** [comment]
Hi @zucchini-nlp, could you take a look at #38331? It's related to the previous PRs of VLM changes. 

I believe `config.text_config` should be read instead of `config` during the `lm_head` initialization in `ForConditionalGeneration` for qwen2_vl and qwen2_5_vl.

File: src/transformers/models/idefics3/modeling_idefics3.py [hunk 3]
```diff
@@ -807,9 +806,6 @@ def forward(
                 past_key_values = DynamicCache()
             past_seen_tokens = past_key_values.get_seq_length()
 
-        if inputs_embeds is not None and input_ids is None and past_seen_tokens == 0:
-            raise ValueError("When first calling the model, if input_embeds are passed, input_ids should not be None.")
-
```
Comment (zucchini-nlp) [inline src/transformers/models/idefics3/modeling_idefics3.py:812]: one is free to pass `embeds`, either if text-only or if they merged vision themselves. We can't raise error for that

File: src/transformers/models/persimmon/modeling_persimmon.py [hunk 4]
```diff
@@ -249,27 +275,29 @@ def forward(
...
-            raise ValueError(
-                f"`attn_output` should be of size {(bsz, self.num_heads, q_len, self.head_dim)}, but is"
-                f" {attn_output.size()}"
-            )
+        if self.config._attn_implementation != "eager":
+            if self.config._attn_implementation == "sdpa" and kwargs.get("output_attentions", False):
+                logger.warning_once(
+                    "`torch.nn.functional.scaled_dot_product_attention` does not support `output_attentions=True`. Falling back to "
+                    'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/persimmon/modeling_persimmon.py:285]: no longer neeed, the config takes care of that!

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py [hunk 5]
```diff
@@ -1952,7 +1773,7 @@ def forward(
                 layer_outputs = self._gradient_checkpointing_func(
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py:1773]: let's use GradientCheckpointingLayer!

**Review by ArthurZucker — APPROVED** [review]
Thanks for pushing this!

File: src/transformers/models/mistral3/modeling_mistral3.py [hunk 6]
```diff
@@ -285,6 +285,9 @@ def get_image_features(
             selected_image_feature = torch.cat(hs_pool, dim=-1)
 
         image_features = self.multi_modal_projector(selected_image_feature.squeeze(0), image_sizes)
+        downsample_ratio = self.vision_tower.patch_size * self.config.spatial_merge_size
+        split_sizes = [(height // downsample_ratio) * (width // downsample_ratio) for height, width in image_sizes]
+        image_features = torch.split(image_features.squeeze(0), split_sizes)
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/mistral3/modeling_mistral3.py:290]: it is not totally BC 🙈 and we are breaking it here, so we can return `len(img_feats) == len(input_img)` ...

File: src/transformers/models/qwen2_5_omni/configuration_qwen2_5_omni.py [hunk 7]
```diff
@@ -396,6 +399,16 @@ def __init__(
...
             self.rope_scaling = {"mrope_section": [16, 24, 24], "rope_type": "default", "type": "default"}
 
+        self.layer_types = layer_types
+        if self.layer_types is None:
+            self.layer_types = [
+                "sliding_attention"
+                if self.sliding_window is not None and i >= self.max_window_layers
+                else "full_attention"
+                for i in range(self.num_hidden_layers)
+            ]
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_omni/configuration_qwen2_5_omni.py:409]: I was copying from Qwen2 to be sure we're compatible with new mask API, not sure if this is planned to be refactored out by Cyril later

**Comment by PeganovAnton** [comment]
@zucchini-nlp @ArthurZucker you need to update `pack_image_features` method docs. Now it returns a list of tensors instead of a single tensor.
